File: js/Pilha.js
const { PilhaCheiaErro, PilhaVaziaErro, TipoErro } = require('./Excecoes');

class Pilha {
    /**
     * @param {number} capacidade - Capacidade máxima da pilha.
     * @param {Function} TypedArrayConstr - Construtor de TypedArray (ex: Int32Array, Float64Array).
     */
    constructor(capacidade, TypedArrayConstr = Int32Array) {
        if (capacidade <= 0) {
            throw new Error("Capacidade deve ser maior que zero.");
        }
        this._capacidade = capacidade;
        this._dados = new TypedArrayConstr(capacidade);
        this._topo = 0; // Quantidade atual de elementos
    }

    empilha(dado) {
        if (this.pilha_esta_cheia()) {
            throw new PilhaCheiaErro("A pilha atingiu a sua capacidade máxima.");
        }

        if (typeof dado !== 'number' || Number.isNaN(dado)) {
            throw new TipoErro(`Dado inválido ou tipo incompatível: ${dado}`);
        }

        this._dados[this._topo] = dado;
        this._topo++;
    }

    desempilha() {
        if (this.pilha_esta_vazia()) {
            throw new PilhaVaziaErro("A pilha está vazia.");
        }

        this._topo--;
        const elemento = this._dados[this._topo];
        return elemento;
    }

    pilha_esta_vazia() {
        return this._topo === 0;
    }

    pilha_esta_cheia() {
        return this._topo === this._capacidade;
    }

    troca() {
        if (this._topo < 2) {
            throw new PilhaVaziaErro("Pilha precisa de pelo menos 2 elementos para efetuar a troca.");
        }

        const temp = this._dados[this._topo - 1];
        this._dados[this._topo - 1] = this._dados[this._topo - 2];
        this._dados[this._topo - 2] = temp;
    }

    tamanho() {
        return this._topo;
    }
}

module.exports = Pilha;

File: cpp/Pilha.cpp
#ifndef PILHA_CPP
#define PILHA_CPP

#include "Pilha.hpp"

template <typename T>
Pilha<T>::Pilha(std::size_t cap) : capacidade(cap), topo(0) {
    if (cap == 0) {
        throw TipoErro("A capacidade deve ser maior que zero.");
    }
    dados = new T[capacidade];
}

template <typename T>
Pilha<T>::~Pilha() {
    delete[] dados;
}

template <typename T>
void Pilha<T>::empilha(T dado) {
    if (pilha_esta_cheia()) {
        throw PilhaCheiaErro("A pilha atingiu a capacidade máxima.");
    }
    dados[topo++] = dado;
}

template <typename T>
T Pilha<T>::desempilha() {
    if (pilha_esta_vazia()) {
        throw PilhaVaziaErro("A pilha está vazia.");
    }
    return dados[--topo];
}

template <typename T>
bool Pilha<T>::pilha_esta_vazia() const {
    return topo == 0;
}

template <typename T>
bool Pilha<T>::pilha_esta_cheia() const {
    return topo == capacidade;
}

template <typename T>
void Pilha<T>::troca() {
    if (topo < 2) {
        throw PilhaVaziaErro("A pilha necessita de ao menos dois elementos para efetuar a troca.");
    }
    T temp = dados[topo - 1];
    dados[topo - 1] = dados[topo - 2];
    dados[topo - 2] = temp;
}

template <typename T>
std::size_t Pilha<T>::tamanho() const {
    return topo;
}

#endif